### DISCUSSION TITLE
OpenFL + Gramine

### DISCUSSION THREAD

**Discussion by lamiaa404** [gramineproject/gramine#1774] [General]
I am trying to replicate the example of using OpenFL with Gramine:[ ](https://github.com/securefederatedai/openfl/blob/develop/openfl-gramine/MANUAL.md). When using gramine-direct the federated learning works without any problems.  However when using gramine-sgx the process just hangs. Any tips how to debug this issue? Attached are two screenshots one with gramine-direct when the federation ends and the other one with gramine-sgx when the process just hangs. 
![Screenshot 2024-02-16 at 15 00 06](https://github.com/gramineproject/gramine/assets/46399141/f249f7d6-bb89-4f58-ad99-a24a23d49c05)
![Screenshot 2024-02-16 at 14 43 35](https://github.com/gramineproject/gramine/assets/46399141/31f9d893-e3be-428f-80c9-5a2d47c37c7d)

**Comment by kailun-qin** [comment]
@lamiaa404: Thanks for the question!

May I know:
1) were you following the [OpenFL+Gramine Manual](https://github.com/securefederatedai/openfl/blob/7a78c055c06d5dcffec1453147ab2fecf2f16833/openfl-gramine/MANUAL.md) and using the [associated artifects](https://github.com/securefederatedai/openfl/tree/7a78c055c06d5dcffec1453147ab2fecf2f16833/openfl-gramine)?
2) which Gramine version were you running w/?

> However when using gramine-sgx the process just hangs

Note that it's mentioned in the [OpenFL+Gramine known issues](https://github.com/securefederatedai/openfl/blob/7a78c055c06d5dcffec1453147ab2fecf2f16833/openfl-gramine/MANUAL.md#known-issues): "Kvasir experiment: aggregation takes really long, debug log-level does not show the reason". So are we sure that it hung rather than running too slowly?

> Any tips how to debug this issue?

Well, we could normally check Gramine logs and/or GDB into it. But let's figure out the above questions first.

**Reply by lamiaa404** [reply]
Hey @kailun-qin! Thanks for the quick reply!

I am currently following the mentioned Manual but for the "Torch_cnn_gramine_ready" experiment. After reading your reply I let it run for much longer time and indeed it just taking realllly long time. So it seems it's an issue not only relating to the Kvasir experiment as written in the known issues but is the case for other workspaces too. 

Thanks again for the help.

**Reply by lamiaa404** [reply]
@kailun-qin Hey again, I have another question :) Whenever I run the federation I get this error: `Error: Emulating a raw syscall instruction. This degrades performance, consider patching your application to use Gramine syscall API.`  Any idea what is causing this error and is there any workaround? I read that this error appears for GO applications. Here in the tutorial PyTorch is used. Thanks!

**Comment by kailun-qin** [comment]
> Whenever I run the federation I get this error: Error: Emulating a raw syscall instruction. This degrades performance, consider patching your application to use Gramine syscall API.
>
> I read that this error appears for GO applications.
>
> Any idea what is causing this error

It's not an error but rather a warning -- statically-built go apps would use raw `SYSCALL` instructions which'll lead gramine-sgx to a "trap-then-emulate" slow path (which would thus impact performance). Pls see more details here:
https://github.com/gramineproject/gramine/blob/e3d3aa39c11054dfa3ce1ad2ce715ec24d730f1a/Documentation/devel/features.md?plain=1#L3232-L3244

> is there any workaround?

If the performance caused by ^ is a concern, you can either try https://go.dev/doc/install/gccgo to produce a dynamically-linked go app, or directly patch the go toolchain to create binaries with the Gramine-specific syscall ABI. (I'd suggest the former as a first shot.)
